### DISCUSSION TITLE
How does "Request Coalescing" work?

### DISCUSSION THREAD

**Discussion by lonix1** [ZiggyCreatures/FusionCache#263] [Q&A] [answered]
The docs explain that a cache stampede is avoided using ["Request Coalescing"](https://github.com/ZiggyCreatures/FusionCache/blob/9d7a5852ebf956f1bb21a68034d309a7ec2c8c16/docs/CacheStampede.md#-request-coalescing), but there's no hint how that works.

Does it simply lock the data-loading factory?

**Comment by jodydonetti** [comment, marked as answer]
Hi @lonix1 , yes that is basically it.

[Request Coalescing](https://support.bunny.net/hc/en-us/articles/6762047083922-Understanding-Request-Coalescing) simply means that multiple requests for the same data at the same time (to be more precise: in the same small time window) are handled as logically the same. To do that, usually a lock is involved, and in FusionCache that is the case with one lock per cache key, to not block a request for product 1 when one for product 2 is being executed, and vice versa.

Hope this helps.

**Reply by lonix1** [reply]
Grazie

**Reply by lonix1** [reply]
The features in this library are incredible. I was going to write my own... but found this library. You've already thought of everything!

BTW, .NET9 is going to have a new "Hybrid Cache". Looks interesting.

**Reply by jodydonetti** [reply]
> The features in this library are incredible.

Thanks @lonix1 , after all the effort put into FusionCache words like these mena a lot, so thanks really!

> I was going to write my own... but found this library. You've already thought of everything!

Glad it has been helpful. Oh, and there are other features I'm working on so no, I haven't though of everything everything 😬

> BTW, .NET9 is going to have a new "Hybrid Cache". Looks interesting.

Yeah I'm [collaborating](https://github.com/dotnet/aspnetcore/issues/53255#issuecomment-1944576582) with the team on that, where "collaborating" means I'm sharing my experiences related to FusionCache with the team and give them some suggestions/ideas, and they are doing the actual work 😅 (also, I suggested the name itself).

**Reply by lonix1** [reply]
Will FusionCache be updated to encapsulate HybridCache, or will they be "competitors"?

(But I don't think they can compete with this, feature-wise. The framework classes usually do only the "basics". If they solve the locking/stampede issue it would be nice though.)

**Reply by jodydonetti** [reply]
> Will FusionCache be updated to encapsulate HybridCache, or will they be "competitors"?

Potentially both, in a way.

The new HybridCache from Microsoft will in theory be 2 things: a sort of _lingua franca_ that others (like FusionCache) can implement and provide additional features, and a concrete out-of-the-box implementation ready to be used.

So on one hand the plan is that I'll create (read: [already working on it](https://x.com/jodydonetti/status/1794373067625599057)) an HybridCache implementation that will act as an adapter for FusionCache, so people and other libs may be able to either depend directly on FusionCache to get all the features, or on a common abstraction (the HybridCache abstract class) while using a different implementation (like FusionCache) without missing some extra features.

On the other hand though they may "compete" in the sense that there's the risk of people just going with "the Microsoft way" and be done with it: they would miss on some extra features (fail-safe, soft/hard timeouts, auto-recovery, etc) but this is their choice.

In the end I'll keep evolving FusionCache to give people new features, extra resiliency and basically a choice, while I hope Microsoft will recognize the existence of other libs like mine, maybe even suggesting them in their blog posts or docs.

We'll see 🤞

> (But I don't think they can compete with this, feature-wise. The framework classes usually do only the "basics". If they solve the locking/stampede issue it would be nice though.)

I don't know about future versions of their HybridCache, but their current plan is to support L1/L1+L2, cache stampede protection, compression and eviction by tag (this last one is still to be decided).

FusionCache on the other hand has more resiliency features ([fail-safe](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/FailSafe.md), [soft/hard timeout](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/Timeouts.md), [eager refresh](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/EagerRefresh.md)), more flexible features ([conditional refresh](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/ConditionalRefresh.md), [named caches](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/NamedCaches.md), [adaptive caching](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/AdaptiveCaching.md)), self-healing ([auto-recovery](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/AutoRecovery.md)) and more so their current feature set is admittedly smaller... for now.

Btw note that I'm also experimenting with compression, and I'm actively working on eviction by tag, we'll see...

Hope I've been able to give you an overall picture.

**Reply by lonix1** [reply]
That was a great answer, thank you. One day you should copy-paste that into the repo's README, because it's very useful!

Seems to me it will be like the difference between `ILogger` vs `Serilog`, `IServiceProvider` vs `Autofac`, `DataAnnotations` vs `FluentValidation` or the new messaging classes vs `Mediatr`. There is room for both. Also in my experience, once they release a full new feature, they rarely add to it, maybe every few years. Whereas an independent library can be evolved quickly. So don't let it stop you.

Having lots of fun using FusionCache so far, many thanks!

**Reply by jodydonetti** [reply]
Hi @lonix1

> That was a great answer, thank you. One day you should copy-paste that into the repo's README, because it's very useful!

Agree, and I kinda did that, by answering [here](https://github.com/ZiggyCreatures/FusionCache/discussions/266#discussioncomment-9915972): I put a lot of details and tried to clarify some nuance, if you have some time let me know what you think.

I think I'll also link that in the main README to be easily discoverable for eveyone wondering about it.

> So don't let it stop you.

Rest assured, I will not 😬
I just hope more people would know about FusionCache, not so much because I've spent a lot of time creating and evolving it and I want some sort of "return" (I don't even accept donations) but because of the underlying reason why I created it in the first place: to help others avoid some common mistakes we all make when caching, to avoid resiliency problems we all run into when in high traffic production environments and to have a ready-made solution that is battle-tested, fast and robust.
Let''s hope!

> Having lots of fun using FusionCache so far, many thanks!

Thanks, happy you're liking it!
